File: src/KRATE.Core/Tools/Images.cs
using System.Buffers.Binary;
using System.Globalization;

namespace Krate.Core;

/// <summary>Image facts read straight from the file header — no decoder, so it stays in pure Core
/// and costs nothing to run. Reading pixels (convert, resize, EXIF) needs a platform imaging API
/// and lives outside Core.</summary>
public static class Images
{
    /// <summary>Format and dimensions of an image, plus megapixels and reduced aspect ratio.</summary>
    public static string Dimensions(string input)
    {
        var path = input.Trim().Trim('"');
        if (!File.Exists(path)) throw new ArgumentException(Strings.Get("Error_NoFile", path));

        // 64 bytes covers every header we read; JPEG needs more, handled inside Read.
        var (format, width, height) = Read(path);
        var mp = width * (long)height / 1_000_000.0;
        var (rw, rh) = ReduceRatio(width, height);
        var named = RatioName(rw, rh);

        return string.Join('\n',
            $"{Strings.Get("Images_Format")}  {format}",
            $"{Strings.Get("Images_Size")}  {width} × {height} px",
            string.Create(CultureInfo.InvariantCulture, $"{Strings.Get("Images_Pixels")}  {mp:0.##} MP"),
            $"{Strings.Get("Images_Ratio")}  {rw}:{rh}{(named is null ? "" : $" ({named})")}",
            $"{Strings.Get("Images_Orientation")}  {Strings.Get(width > height ? "Images_Landscape" : width < height ? "Images_Portrait" : "Images_Square")}");
    }

    /// <summary>Reads (format, width, height) from the magic bytes. Throws if unrecognised.</summary>
    public static (string Format, int Width, int Height) Read(string path)
    {
        using var stream = File.OpenRead(path);
        var head = new byte[Math.Min(32, stream.Length)];
        stream.ReadExactly(head, 0, head.Length);
        var b = head.AsSpan();

        // PNG: 8-byte signature, then IHDR with big-endian width/height at offset 16.
        if (b.Length >= 24 && b[..8].SequenceEqual(PngSignature))
            return ("PNG", ReadBE32(b[16..]), ReadBE32(b[20..]));

        // GIF: "GIF87a"/"GIF89a", little-endian width/height at offset 6.
        if (b.Length >= 10 && b[..3].SequenceEqual("GIF"u8))
            return ("GIF", BinaryPrimitives.ReadUInt16LittleEndian(b[6..]), BinaryPrimitives.ReadUInt16LittleEndian(b[8..]));

        // BMP: "BM", signed little-endian width/height at offset 18 (height can be negative = top-down).
        if (b.Length >= 26 && b[0] == 'B' && b[1] == 'M')
            return ("BMP", BinaryPrimitives.ReadInt32LittleEndian(b[18..]), Math.Abs(BinaryPrimitives.ReadInt32LittleEndian(b[22..])));

        // WebP: RIFF container tagged "WEBP".
        if (b.Length >= 30 && b[..4].SequenceEqual("RIFF"u8) && b[8..12].SequenceEqual("WEBP"u8))
            return Webp(b);

        // JPEG: FF D8, then walk the segments to the frame header.
        if (b.Length >= 2 && b[0] == 0xFF && b[1] == 0xD8)
            return Jpeg(stream);

        throw new ArgumentException(Strings.Get("Error_UnknownImage", Path.GetFileName(path)));
    }

    static (string, int, int) Webp(ReadOnlySpan<byte> b)
    {
        var chunk = b[12..16];
        if (chunk.SequenceEqual("VP8X"u8)) // extended: 24-bit canvas size, stored as value-1
            return ("WebP", 1 + Read24LE(b[24..]), 1 + Read24LE(b[27..]));
        if (chunk.SequenceEqual("VP8L"u8)) // lossless: 14 bits each after the 0x2F signature
        {
            var bits = BinaryPrimitives.ReadUInt32LittleEndian(b[21..]);
            return ("WebP", 1 + (int)(bits & 0x3FFF), 1 + (int)(bits >> 14 & 0x3FFF));
        }
        if (chunk.SequenceEqual("VP8 "u8)) // lossy: 14-bit dimensions after the 0x9D012A start code
            return ("WebP", BinaryPrimitives.ReadUInt16LittleEndian(b[26..]) & 0x3FFF,
                            BinaryPrimitives.ReadUInt16LittleEndian(b[28..]) & 0x3FFF);
        throw new ArgumentException("WebP");
    }

    static (string, int, int) Jpeg(FileStream stream)
    {
        stream.Position = 2;
        Span<byte> marker = stackalloc byte[4];
        while (stream.Read(marker[..2]) == 2)
        {
            if (marker[0] != 0xFF) throw new ArgumentException("JPEG");
            var kind = marker[1];
            // SOF0-3, 5-7, 9-11, 13-15 carry the dimensions; everything else we skip by its length.
            if (kind is >= 0xC0 and <= 0xCF && kind is not (0xC4 or 0xC8 or 0xCC))
            {
                stream.Position += 3; // length(2) + precision(1)
                stream.ReadExactly(marker);
                return ("JPEG", BinaryPrimitives.ReadUInt16BigEndian(marker[2..]), BinaryPrimitives.ReadUInt16BigEndian(marker[..2]));
            }
            if (kind is 0xD8 or 0xD9 or (>= 0xD0 and <= 0xD7)) continue; // markers without a length
            stream.ReadExactly(marker[..2]);
            stream.Position += BinaryPrimitives.ReadUInt16BigEndian(marker[..2]) - 2;
        }
        throw new ArgumentException("JPEG");
    }

    /// <summary>"1920x1080" → its reduced ratio; "16:9 1920" → the matching height.</summary>
    public static string Ratio(string input)
    {
        var s = input.Trim().ToLowerInvariant();
        var parts = s.Split([' ', ','], StringSplitOptions.RemoveEmptyEntries);

        // "16:9 1920" — fill in the missing dimension from a known one.
        if (parts.Length == 2 && parts[0].Contains(':') && int.TryParse(parts[1], out var known))
        {
            var ratio = parts[0].Split(':');
            var (rw, rh) = (int.Parse(ratio[0], CultureInfo.InvariantCulture), int.Parse(ratio[1], CultureInfo.InvariantCulture));
            return string.Join('\n',
                $"{known} × {(long)known * rh / rw} px  ({Strings.Get("Images_FromWidth")})",
                $"{(long)known * rw / rh} × {known} px  ({Strings.Get("Images_FromHeight")})");
        }

        // "1920x1080" — reduce it.
        var wh = s.Split(['x', ':', '×'], StringSplitOptions.RemoveEmptyEntries);
        if (wh.Length != 2 || !int.TryParse(wh[0], out var w) || !int.TryParse(wh[1], out var h) || w <= 0 || h <= 0)
            throw new ArgumentException(Strings.Get("Error_RatioUsage"));
        var (a, c) = ReduceRatio(w, h);
        var name = RatioName(a, c);
        return $"{a}:{c}{(name is null ? "" : $"  ({name})")}";
    }

    static (int, int) ReduceRatio(int w, int h)
    {
        var g = (int)Maths.Gcd(w, h);
        return g == 0 ? (w, h) : (w / g, h / g);
    }

    // 16:10 reduces to 8:5, so both spellings map to the same name.
    static string? RatioName(int w, int h) => (w, h) switch
    {
        (16, 9) => "16:9",
        (4, 3) => "4:3",
        (3, 2) => "3:2",
        (8, 5) => "16:10",
        (21, 9) or (7, 3) => "21:9",
        (1, 1) => "1:1",
        (5, 4) => "5:4",
        _ => null,
    };

    static ReadOnlySpan<byte> PngSignature => [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A];

    static int ReadBE32(ReadOnlySpan<byte> b) => (int)BinaryPrimitives.ReadUInt32BigEndian(b);
    static int Read24LE(ReadOnlySpan<byte> b) => b[0] | b[1] << 8 | b[2] << 16;
}

File: rust/src/images.rs
//! Image dimensions read from magic bytes. Mirrors `Krate.Core.Images.Read`/`Dimensions`.
//!
//! No image library involved: every format here declares its size in the first few bytes, so
//! this is byte parsing rather than decoding. That is also why it ports exactly — an `image`
//! crate would give the same numbers but is a dependency for nothing.

use crate::convert::{ratio_name, reduce_ratio};
use crate::i18n;
use crate::tools::format_decimal;

const PNG_SIGNATURE: [u8; 8] = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A];

fn be32(b: &[u8]) -> i64 {
    i64::from(u32::from_be_bytes([b[0], b[1], b[2], b[3]]))
}

fn le16(b: &[u8]) -> i64 {
    i64::from(u16::from_le_bytes([b[0], b[1]]))
}

fn le32(b: &[u8]) -> i64 {
    i64::from(i32::from_le_bytes([b[0], b[1], b[2], b[3]]))
}

fn le24(b: &[u8]) -> i64 {
    i64::from(u32::from_le_bytes([b[0], b[1], b[2], 0]))
}

/// Reads (format, width, height) from the magic bytes.
pub fn read(bytes: &[u8], name: &str) -> Result<(&'static str, i64, i64), String> {
    let unknown = || i18n::format("Error_UnknownImage", &[name]);

    // PNG: 8-byte signature, then IHDR with big-endian width/height at offset 16.
    if bytes.len() >= 24 && bytes[..8] == PNG_SIGNATURE {
        return Ok(("PNG", be32(&bytes[16..]), be32(&bytes[20..])));
    }
    // GIF: "GIF87a"/"GIF89a", little-endian width/height at offset 6.
    if bytes.len() >= 10 && &bytes[..3] == b"GIF" {
        return Ok(("GIF", le16(&bytes[6..]), le16(&bytes[8..])));
    }
    // BMP: "BM", signed little-endian at offset 18; a negative height means top-down.
    if bytes.len() >= 26 && bytes[0] == b'B' && bytes[1] == b'M' {
        return Ok(("BMP", le32(&bytes[18..]), le32(&bytes[22..]).abs()));
    }
    // WebP: RIFF container tagged "WEBP".
    if bytes.len() >= 30 && &bytes[..4] == b"RIFF" && &bytes[8..12] == b"WEBP" {
        return webp(bytes);
    }
    // JPEG: FF D8, then walk the segments to the frame header.
    if bytes.len() >= 2 && bytes[0] == 0xFF && bytes[1] == 0xD8 {
        return jpeg(bytes);
    }
    Err(unknown())
}

fn webp(b: &[u8]) -> Result<(&'static str, i64, i64), String> {
    let chunk = &b[12..16];
    if chunk == b"VP8X" {
        // Extended: 24-bit canvas size, stored as value-1.
        return Ok(("WebP", 1 + le24(&b[24..]), 1 + le24(&b[27..])));
    }
    if chunk == b"VP8L" {
        // Lossless: 14 bits each after the 0x2F signature.
        let bits = u32::from_le_bytes([b[21], b[22], b[23], b[24]]);
        return Ok((
            "WebP",
            1 + i64::from(bits & 0x3FFF),
            1 + i64::from((bits >> 14) & 0x3FFF),
        ));
    }
    if chunk == b"VP8 " {
        // Lossy: 14-bit dimensions after the 0x9D012A start code.
        return Ok(("WebP", le16(&b[26..]) & 0x3FFF, le16(&b[28..]) & 0x3FFF));
    }
    Err("WebP".to_string())
}

/// Walks the segment chain to the start-of-frame header, which is the only place the dimensions
/// live. Segments are skipped by their declared length.
fn jpeg(bytes: &[u8]) -> Result<(&'static str, i64, i64), String> {
    let bad = || "JPEG".to_string();
    let mut pos = 2usize;

    while pos + 1 < bytes.len() {
        if bytes[pos] != 0xFF {
            return Err(bad());
        }
        let kind = bytes[pos + 1];
        pos += 2;

        // SOF0-3, 5-7, 9-11, 13-15 carry the dimensions; 0xC4/0xC8/0xCC do not.
        if (0xC0..=0xCF).contains(&kind) && !matches!(kind, 0xC4 | 0xC8 | 0xCC) {
            // length(2) + precision(1), then height then width, both big-endian.
            let at = pos + 3;
            if at + 4 > bytes.len() {
                return Err(bad());
            }
            let height = i64::from(u16::from_be_bytes([bytes[at], bytes[at + 1]]));
            let width = i64::from(u16::from_be_bytes([bytes[at + 2], bytes[at + 3]]));
            return Ok(("JPEG", width, height));
        }
        // Markers that carry no length payload.
        if matches!(kind, 0xD8 | 0xD9) || (0xD0..=0xD7).contains(&kind) {
            continue;
        }
        if pos + 2 > bytes.len() {
            return Err(bad());
        }
        let length = usize::from(u16::from_be_bytes([bytes[pos], bytes[pos + 1]]));
        if length < 2 {
            return Err(bad());
        }
        pos += length;
    }
    Err(bad())
}

/// Format and dimensions of an image, plus megapixels and reduced aspect ratio.
pub fn dimensions(input: &str) -> Result<String, String> {
    let path = input.trim().trim_matches('"');
    let bytes = std::fs::read(path).map_err(|_| i18n::format("Error_NoFile", &[path]))?;
    let name = std::path::Path::new(path)
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| path.to_string());

    let (format, width, height) = read(&bytes, &name)?;
    let megapixels = width as f64 * height as f64 / 1_000_000.0;
    let (rw, rh) = reduce_ratio(width, height);
    let named = ratio_name(rw, rh);

    let orientation = if width > height {
        "Images_Landscape"
    } else if width < height {
        "Images_Portrait"
    } else {
        "Images_Square"
    };

    Ok([
        format!("{}  {format}", i18n::get("Images_Format")),
        format!("{}  {width} × {height} px", i18n::get("Images_Size")),
        format!("{}  {} MP", i18n::get("Images_Pixels"), format_decimal(megapixels, 2)),
        format!(
            "{}  {rw}:{rh}{}",
            i18n::get("Images_Ratio"),
            named.map(|n| format!(" ({n})")).unwrap_or_default()
        ),
        format!("{}  {}", i18n::get("Images_Orientation"), i18n::get(orientation)),
    ]
    .join("\n"))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Headers are hand-built, exactly as the C# test suite does — the magic bytes and offsets
    /// are what a real file carries, so no image files or decoder are needed.
    fn png(w: u32, h: u32) -> Vec<u8> {
        let mut b = vec![0u8; 24];
        b[..8].copy_from_slice(&PNG_SIGNATURE);
        b[16..20].copy_from_slice(&w.to_be_bytes());
        b[20..24].copy_from_slice(&h.to_be_bytes());
        b
    }

    fn gif(w: u16, h: u16) -> Vec<u8> {
        let mut b = vec![0u8; 10];
        b[..6].copy_from_slice(b"GIF89a");
        b[6..8].copy_from_slice(&w.to_le_bytes());
        b[8..10].copy_from_slice(&h.to_le_bytes());
        b
    }

    fn bmp(w: i32, h: i32) -> Vec<u8> {
        let mut b = vec![0u8; 26];
        b[0] = b'B';
        b[1] = b'M';
        b[18..22].copy_from_slice(&w.to_le_bytes());
        b[22..26].copy_from_slice(&h.to_le_bytes());
        b
    }

    fn jpeg_bytes(w: u16, h: u16) -> Vec<u8> {
        // SOI, then a dummy APP0 with a length, then SOF0 carrying the dimensions.
        let mut b = vec![0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x04, 0x00, 0x00, 0xFF, 0xC0, 0x00, 0x11, 0x08];
        b.extend_from_slice(&h.to_be_bytes());
        b.extend_from_slice(&w.to_be_bytes());
        b
    }

    #[test]
    fn reads_every_format_from_its_header() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        assert_eq!(read(&png(1920, 1080), "x").unwrap(), ("PNG", 1920, 1080));
        assert_eq!(read(&gif(320, 240), "x").unwrap(), ("GIF", 320, 240));
        assert_eq!(read(&bmp(640, 480), "x").unwrap(), ("BMP", 640, 480));
        assert_eq!(read(&jpeg_bytes(800, 600), "x").unwrap(), ("JPEG", 800, 600));
    }

    /// A BMP stores height negative when the rows run top-down; the magnitude is the height.
    #[test]
    fn bmp_top_down_height_is_absolute() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        assert_eq!(read(&bmp(640, -480), "x").unwrap(), ("BMP", 640, 480));
    }

    /// JPEG dimensions are height-then-width, and the parser must skip other segments to find
    /// the frame header rather than reading the first thing after SOI.
    #[test]
    fn jpeg_walks_past_other_segments() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        let (_, w, h) = read(&jpeg_bytes(1234, 567), "x").unwrap();
        assert_eq!((w, h), (1234, 567), "width and height must not be swapped");
    }

    #[test]
    fn unrecognised_and_truncated_input_is_rejected() {
        let _guard = crate::i18n::test_lock();
        i18n::set_language("en");
        assert!(read(b"not an image at all, really", "x.txt").is_err());
        assert!(read(&[], "empty").is_err());
        assert!(read(&PNG_SIGNATURE, "truncated.png").is_err(), "signature but no IHDR");
        // A JPEG that never reaches a frame header must fail rather than loop.
        assert!(read(&[0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x04, 0x00, 0x00], "x.jpg").is_err());
    }
}
